### ISSUE TITLE
feature request: pick maximum file size instead of bitrate

### ISSUE THREAD

**Issue by codl** [brarcher/video-transcoder#56]
often when I'm transcoding a video it is to make it fit within a file size limit, and i end up having to do some math to figure out what is the best bitrate i can set while staying under that size. I'd like the ability to pick a maximum size in MB and have video transcoder pick (or suggest) an appropriate bitrate

**Comment by phuyuk** [comment]
This would be very useful since, for me, the main purpose of this app is to shrink a video down to an appropriate size for email. However, directly fiddling with the bitrate has its place, too.

Ideally, if setting one variable, the other should be calculated/estimated and shown. For example, if I was to stipulate a 10MB filesize for a 20 minute clip, it would be useful to see that the corresponding bitrate is 20 kb/s and the resulting quality will be unacceptable. I might then elect to split the file before re-encoding.

Anyway, it's all well and good to make these requests - I'm not the one who has to implement them. The app already has utility and I'm grateful for it. Thank you.

**Comment by licaon-kter** [comment]
How do you profile this exactly? You can't it depends on content too.

You could try to encode a segment, say 5 seconds, see its size, extrapole to the whole lenght of time and decide to keep the settings or try newer ones.

Now that I've written this I realised that's a nice way for the app to do it...automatically 🤣

**Comment by phuyuk** [comment]
Cool!  I think a rough approximation would be sufficient for my purposes. But, if you're going down that road... intuitively, it seems preferential to take a distribution of tiny samples, across the entire video to estimate the content. Let's say the single, 5-second contiguous sample lies somewhere in the intro of my new art film (lol); the first minute of the clip is a climactic scene from a Swedish drama (ie. a still shot of a dark room and a chair), followed by 9 minutes of daytime POV rollercoaster terror, in a field of sunflowers and tulips. Maybe even ignore the first and last 10%, since finished videos usually have chairs at both ends. ;)

**Comment by codl** [comment]
> directly fiddling with the bitrate has its place, too

agreed. i meant this as an *alternate setting* to supplying your own bitrate, i did not mean to replace that entirely

---

> How do you profile this exactly? You can't it depends on content too.
>
> You could try to encode a segment, say 5 seconds

> take a distribution of tiny samples, across the entire video

yall... you're overthinking this. video file size isn't that unpredictable

here's a crash course in rate control modes in audio / video encoding, and how they would apply here

* constant bitrate

  the size of a constant bitrate media is literally dictated by the bitrate you tell it to use: 

  `size = time in seconds × (video bitrate + audio bitrate + some small overhead which depends on the container format)`

  in practice i've found the container overhead to be negligible but, to be safe you can say it's going to be less than 1kbps and just use 1k in that equation

  which, given the desired size, and the video time, would give us `video bitrate + audio bitrate = (size / time) - 1000`

  let the user pick a audio bitrate, and now you know which video bitrate to use :)

* average bitrate

  this is the one you get by default when you use `-b` in ffmpeg. this is also what this app uses currently [\[1\]](https://github.com/brarcher/video-transcoder/blob/34b4d8c00959e3e173fe9e80aca6a71c88f969b6/app/src/main/java/protect/videotranscoder/activity/MainActivity.java#L454-L455)

  in practice, average bitrate videos hover around the same size as constant bitrate with the same settings, because the bitrate averaged over time is the same. so we can reuse the same equation and get the same results!

  you can't be certain that the video will be *under* the target size however, it could overshoot a little if the last few seconds go over the average, but you can always nudge the file size down by a couple MB without telling the user 👀

  also you can force ffmpeg to never go over by setting both `-maxrate` and `-bufsize` to the bitrate you want

* variable bitrate (also known as constant rate factor)

  that's the unpredictable one. don't use that one for this. don't use that one in general for video that's gonna be streamed over a network. it's excellent for archival but that's probably not what you're doing if you're transcoding a video on your phone

  this is probably the one yall were thinking of

**Comment by phuyuk** [comment]
"... it's excellent for archival but that's probably not what you're doing if you're transcoding a video on your phone."

No, that's exactly what I'm doing.

With all due respect, emailing is not streaming, and variable bitrate (what I'd assumed we were discussing, all along) should yield a smaller file. So, that's an option I'd vote for - that is, if it's not too much trouble to implement. Options are good. And this is a good option... IMHO.

**Comment by codl** [comment]
then make your own issue for VBR?

**Comment by codl** [comment]
sorry, I didn't mean to start a control rate argument. i was just trying to explain why it was fairly easy to get a controlled file size when using CBR or ABR, and how to do it. i only mentioned VBR for completeness's sake, as the app doesn't let you use VBR anyway right now so it's not really relevant to this issue

**Comment by phuyuk** [comment]
"the app doesn't let you use VBR anyway"  As I recall, FFmpeg has a fairly complete VBR implementation. At any rate (lol), the feature you've requested naturally belongs in this app, and I look forward to it, with or without VBR.

Sorry about hijacking your thread.
